mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapperTest.java

```diff
@@ -353,4 +356,63 @@ void addShouldRetryMessageDenormalization(CassandraCluster cassandra) throws Exc
                 .hasSize(1);
         }
     }
+
+    @Nested
+    class ReadRepairsTesting {
+        @Test
+        void readingShouldEventuallyFixCountersInconsistencies(CassandraCluster cassandra) throws MailboxException {
+            saveMessages();
+            FlagsUpdateCalculator newFlags = new FlagsUpdateCalculator(new Flags(Flags.Flag.SEEN), MessageManager.FlagsUpdateMode.REPLACE);
+            messageMapper.updateFlags(benwaInboxMailbox, message1.getUid(), newFlags);
+            // Expected count of unseen is 4 see MessageMapperTest::mailboxUnSeenCountShouldBeDecrementedAfterAMessageIsMarkedSeen
+
+            // Create an inconsistency
+            new CassandraMailboxCounterDAO(cassandra.getConf())
+                .incrementUnseenAndCount((CassandraId) benwaInboxMailbox.getMailboxId())
+                .block();
+
+            // 100 poll with a 0.1 probability to trigger read repair
+            Awaitility.await()
+                .pollInterval(new Duration(10, MILLISECONDS))
+                .atMost(Duration.ONE_SECOND)
+                .untilAsserted(() ->
+                    assertThat(messageMapper.getMailboxCounters(benwaInboxMailbox).getUnseen()).isEqualTo(4));
+        }
+
+        @Test
+        void readingShouldEventuallyFixMissingCountersInconsistencies(CassandraCluster cassandra) throws MailboxException {
+            saveMessages();
+            FlagsUpdateCalculator newFlags = new FlagsUpdateCalculator(new Flags(Flags.Flag.SEEN), MessageManager.FlagsUpdateMode.REPLACE);
+            messageMapper.updateFlags(benwaInboxMailbox, message1.getUid(), newFlags);
+            // Expected count of unseen is 4 see MessageMapperTest::mailboxUnSeenCountShouldBeDecrementedAfterAMessageIsMarkedSeen
+
+            // Create an inconsistency
+            new CassandraMailboxCounterDAO(cassandra.getConf())
+                .delete((CassandraId) benwaInboxMailbox.getMailboxId())
+                .block();
+
+            // 100 poll with a 0.1 probability to trigger read repair
+            Awaitility.await()
+                .pollInterval(new Duration(10, MILLISECONDS))
+                .atMost(Duration.ONE_SECOND)
+                .untilAsserted(() ->
+                    assertThat(messageMapper.getMailboxCounters(benwaInboxMailbox).getUnseen()).isEqualTo(4));
+        }
+
+        @Test
+        void readingShouldFixInvalidCounters(CassandraCluster cassandra) throws MailboxException {
```

Comment:
I'm not sure to get it between this test and the first one `readingShouldEventuallyFixCountersInconsistencies`

`readingShouldEventuallyFixCountersInconsistencies` => we create an inconsistency, do 100 reads with the fact that it does not happen everytime and check it got solved
`readingShouldFixInvalidCounters` => we create an inconsistency, one read and it's fixed? 

Or did I miss something?

Reply:
If the counter is invalid, we know for sure we have a consistency issue, we can fix it straight away, and return **directly** the right answer.

Probabilistic read repairs are probabilistic: when the data looks OK, we have say 1 chance out of 10, to perform a repair.